### PR TITLE
Optional Pagination

### PR DESCRIPTION
**Pull request by fruitymedley**
Implements optional pagination as requested in #167 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

#167 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)

**Comment by fruitymedley** [comment]
@bellini666 Hi! Wanted to flag this to you for review since you look like the most recent active contributor, or if you know who would be best, I would greatly appreciate it!

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (bellini666) [inline src/strawberry_sqlalchemy_mapper/relay.py:149]: I wonder if this would be better defined in `mapper.type` as an argument?

Also, the name itself IMO implies you are excluding relay functionality, which is not totally true. My suggestion would be do the argument and call it `map_relations_as_list: bool = True`, in which you can opt-out as `False` (would be even better to be the other way around, but then it would be a breaking change)

**Review by bellini666 — COMMENTED** [review]
The solution in general looks good. I would just change it to be an argument instead, so it is more aligned to how we do those kind of configurations on strawberry.

Maybe @patrick91 and/or @erikwrede can also take a look at this?

File: src/strawberry_sqlalchemy_mapper/relay.py
Comment (fruitymedley) [reply src/strawberry_sqlalchemy_mapper/relay.py:149]: I could see a situation in which a type might have both relay and list. Would it make sense to add a class property in the same vein as `__exclude__`?

**Comment by fruitymedley** [comment]
I've made a few updates in accordance with the suggestions. Please review as available.

**Comment by patrick91** [comment]
@fruitymedley would this work for you as a global configuration, instead of a configuration on the type itself?

**Comment by novag** [comment]
@fruitymedley Now that the project has a maintainer again in @Ckk3, perhaps we can get this merged. @Ckk3 What do you think about the global vs typed-based configuration?

**Comment by erikwrede** [comment]
Sorry, I'm a bit late to the party... 

I agree with Patrick about having a global config on how to handle `*-n` relationships. Having a class attribute seems to be an easy fix, but I'm sceptical about the devX (no linting on field renames or other errors, the schema type will just change)

Another, perhaps more opinionated approach, would be to work with the global config and use an explicit field definition, like `doNotUseRelayField: list[MyType]` , or even list[auto] on each field that doesn't follow the global config. This is certainly a different approach, but I prefer that it's in line with the declarative style of other strawberry fields.

**Review by Ckk3 — CHANGES_REQUESTED** [review]
Thanks a lot, @fruitymedley , for your contribution! 🙌 This PR is looking great, and I just have a minor change related to the tests. I haven’t made any changes to the PR itself, but I wanted to note that if think we’ll need to update the documentation later to reflect the new functionality.

I'll also send a separate message soon with my thoughts on the global vs. type-based configuration discussion.

Looking forward to your feedback on the test updates!

**Comment by Ckk3** [comment]
> @fruitymedley Now that the project has a maintainer again in @Ckk3, perhaps we can get this merged. @Ckk3 What do you think about the global vs typed-based configuration?

Hi everyone!

I don’t have a strong preference, but after reviewing the code and looking at the implementation in Strawberry Django, I think the current approach in this PR solves the issue. However, I believe a global configuration might be a better solution.

_Please correct me if I’m wrong, @bellini666 , as I know you’re more familiar with the Django integration._

In Strawberry Django’s documentation, I noticed that pagination can be configured both at the type level and on individual fields.

- In the [Pagination Docs](https://strawberry.rocks/docs/django/guide/pagination), we see pagination being set using `pagination=True` inside `@strawberry_django.type`.
image

- Then, in the [Fields Docs](https://strawberry.rocks/docs/django/guide/fields), under the "Field Customization" section, pagination is set on a specific field within a type.
image

If I understand correctly, this PR introduces the second approach (configuring pagination on individual fields within a type), but I agree with @erikwrede's point:
> devX (no linting on field renames or other errors, the schema type will just change)

I also align with @bellini666  later comment. In my opinion, we should apply this configuration at the type level, similar to `@strawberry_django.type(pagination=True)`, as it simplifies development. Afterward, we could discuss adding an option to configure it on individual fields if needed.

What do you all think? Please let me know if I misunderstood anything, I'm still getting familiar with the Django integration.

**Comment by Ckk3** [comment]
Hi, @fruitymedley  , we have fixed the CI, can you please update your branch from main?

**Comment by bellini666** [comment]
@Ckk3 @erikwrede so, the way Strawberry Django currently does is to enforce a declarative style, in which you can use `auto` or properly [annotate the relation](https://strawberry.rocks/docs/django/guide/fields#relationships).

For example:

```python
@strawberry_django.type(Fruit):
class FruitType:
    ...


@strawberry_django.type(Color):
class ColorType:
    # Does not auto resolve to the type, as we don't keep a mapping internally.
    # Instead, one-to-one/many-to-one relations are mapped to `DjangoModelType` and
    # one-to-many/many-to-many are mapped to `list[DjangoModelType]`
    # `DjangoModelType` is a simple type that contains `pk` in it:
    # https://github.com/strawberry-graphql/strawberry-django/blob/main/strawberry_django/fields/types.py#L75
    fruits: auto

    # The recommended way for lists. The integration will still know that we have a
    # relation in the model and do what needs to be done regarding optimization
    fruits: list[Fruit]

    # The same as list above, but use relay connections instead.
    fruits: relay.ListConnection[Fruit]
```

I know that strawberry-sqlalchemy currently doesn't support the declarative style very well. Hence, a solution like the one proposed here is probably the way to go for now. Still, I'd love to see this integration moving forward to now only allow, but maybe actually enforce that in the future, so it is more aligned to strawberry style and recommendations in general.

**Comment by Ckk3** [comment]
I agree with @bellini666 , lets continue with this PR when you can @fruitymedley 
I also hope we can improve this lib later to make declarative style easier to develop.

**Comment by Ckk3** [comment]
Hi @fruitymedley,

This PR has been inactive for two weeks. Are you still able to work on it? If not, please let me know so we can continue the development from this point.

**Comment by Ckk3** [comment]
I will continue the development of this PR.

**Comment by Ckk3** [comment]
I've fixed this PR tests and update the code with the new strawberry version.
Please take a look again when you can @patrick91 @bellini666 !

**Review by bellini666 — APPROVED** [review]

**Review by Ckk3 — APPROVED** [review]

**Comment by Ckk3** [comment]
I'll merge this PR.
Thanks @fruitymedley for your work!

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**
